confd/pkg/backends/calico/client.go

```diff
@@ -1086,6 +1087,31 @@ func (c *client) getListenPortKVPair(v3res *apiv3.BGPConfiguration, key interfac
 	*updatePeersV1 = true
 }
 
+func (c *client) getBindAddressKVPair(v3res *apiv3.BGPConfiguration, key interface{}, updatePeersV1 *bool, updateReasons *[]string) {
+	bindAddrV4 := getBGPConfigKey("bind_address_v4", key)
+	bindAddrV6 := getBGPConfigKey("bind_address_v6", key)
+
+	if v3res != nil && v3res.Spec.BindMode != "" {
+		if v3res.Spec.BindMode == "None" {
+
+		}
+		if v3res.Spec.BindMode == "NodeIP" {
+
+		}
+		*updateReasons = append(*updateReasons, "bindAddress updated.")
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(bindAddrV4, ""))
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(bindAddrV6, ""))
+	} else {
+		if v3res != nil && v3res.Spec.BindMode == "" {
+
+		}
+		*updateReasons = append(*updateReasons, "bindAddress deleted.")
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(bindAddrV4))
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(bindAddrV6))
+	}
+	*updatePeersV1 = true
```

Comment:
This is obviously missing code as it stands.  But anyway I would suggest just propagating the BindMode value so that the template can see it:
```
bindMode := getBGPConfigKey("bind_mode", key)
c.updateCache(api.UpdateTypeKVUpdated, getKVPair(bindMode, v3res.Spec.BindMode))
```

Then the templating should check for bind_mode in a similar fashion to this:
```
{{- $node_listen_port_key := printf "/host/%s/listen_port" (getenv "NODENAME")}}
{{- if exists $node_listen_port_key}}
```
If it exists and equals NodeIP, emit the `listen bgp` line with `$node_ip` as the address.

Reply:
yes modified accordingly. But when i apply below config, its still not able to recognize `bindMode` . Did i miss anything ?

```

apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  listenPort: 177
  bindMode: NodeIP
```

Reply:
@hanamantagoudvk Can you tell me exactly how you've tested, so that I can think it through?

Reply:
@neiljerram  As per Shaun's suggestion, i built `calicoctl` . With new binary also, i get the same above error.

Reply:
Ah, I think you misunderstood my comment about replacing "bindMode" with "oneof...".  I meant only for the "validate" part of the tag.  So the whole BindMode field should be:
```
	BindMode string `json:"bindMode,omitempty" validate:"omitempty,oneof=None NodeIP"`
```
It's possible there might still be other issues, but you definitely need that one corrected.

Reply:
@neiljerram @fasaxc after modifying `bgpconfig.go` as per suggestion, i rebuilt `typa `and `calicoctl`. After that i am not facing above error issue. I could see that value of bindMode is properly being set in caches. Its also being returned by `GetValues `method. But only thing is , i dont see that its reflecting in `/etc/calico/confd/config/bird.cfg` eventually.

Reply:
I'm afraid you'll need to rebuild confd as well, and the calico-node image if that is what you are using for testing.  Can you tell us more about your test procedure?

Reply:
@neiljerram , yes i am rebuilding confd and calico-node images. Then i am deploying calico-node daemonset using this  image in k8s cluster. Then as i said earlier , i apply the `bgp-cfg.yaml` (see above) which has bindMode set to NodeIP. Then when i check the `netstat -an | grep 179`  on worker nodes, i should see that bird is listening on `<node-ip>:179` . But i dont see that happening . I see `0.0.0.0:179`.

Reply:
I suggest adding some `log.Info` lines into `getBindModeKVPair`, then check the confd logging (which is part of the calico-node log) to see if your logic is really being hit and behaving correctly.